### ISSUE TITLE
Consider adding an event to the OAuth 2.0 base handler to validate token responses

### ISSUE THREAD

**Issue by kevinchalet** [dotnet/aspnetcore#23749]
In some cases, developers want to make sure that the scopes granted by the authorization server/resource owner exactly match the requested scopes. In OAuth 2.0/OIDC, this can be done by determining whether a `scope` property was returned as part of the authorization response (for the implicit flow, but it's easy to bypass) or the token response, and ensuring it matches the requested scopes.

Unfortunately, there's currently no event that would allow implementing that.
Alternatively, if we think it's a common enough scenario, it should be natively implemented as a `ValidateGrantedScope` option.

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/pull/444#issuecomment-655032756

/cc @Tratcher @bryantlikes

**Comment by Tratcher** [comment]
So for OAuth there'd be a change to capture the scopes about here as part of the OAuthTokenResponse:
https://github.com/dotnet/aspnetcore/blob/a5901c084aab5aba6162a57b6fe2dc0ec7764e48/src/Security/Authentication/OAuth/src/OAuthHandler.cs#L204

Then you could check the scopes in the existing CreateTicket event, right?
https://github.com/dotnet/aspnetcore/blob/a5901c084aab5aba6162a57b6fe2dc0ec7764e48/src/Security/Authentication/OAuth/src/OAuthHandler.cs#L165

For OIDC implicit flow could you do this check in the MessageReceived event?
https://github.com/dotnet/aspnetcore/blob/a5901c084aab5aba6162a57b6fe2dc0ec7764e48/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L527

For the OIDC token endpoint response, it seems like you could check this in the TokenResponseReceived event.
https://github.com/dotnet/aspnetcore/blob/a5901c084aab5aba6162a57b6fe2dc0ec7764e48/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L658

**Comment by kevinchalet** [comment]
> So for OAuth there'd be a change to capture the scopes about here as part of the OAuthTokenResponse:

Yep. Now the question is: do we want the base OAuth 2.0 handler to do that or do we just want to provide an event to allow folks to implement that themselves?

> For OIDC implicit flow could you do this check in the MessageReceived event?

Yeah, the OIDC handler has a much better events story and it's already possible to support that scenario using the existing hooks (that's why I only mentioned the OAuth 2.0 handler in my OP).
